### ISSUE TITLE
Improve exception safety with smart pointers

### ISSUE THREAD

**Issue by elfring** [01org/parameter-framework#311]
Would you like to wrap any pointer data members with the template class "[std::unique_ptr](https://en.wikipedia.org/wiki/Smart_pointer#unique_ptr "a smart pointer")"?

Update candidates:
* [CParameterMgr](https://github.com/01org/parameter-framework/blob/fe92bf52e505b8c9b5a2dd063f0007017c23feee/parameter/ParameterMgr.cpp#L336 "CParameterMgr class")
* [CParameterMgrPlatformConnector](https://github.com/01org/parameter-framework/blob/77b32fb5bd192ee240bab33eb733839670671e38/parameter/ParameterMgrPlatformConnector.cpp#L38 "CParameterMgrPlatformConnector class")

**Comment by krocard** [comment]
I agree with you. Nevertheless we do not plan on updating the code to C++11 in one big bang. Rather each time some code is modified, it is updated in the same time. I will mark this enhancement as trivial. Maybe a newcomer will pick it up.

**Comment by divserge** [comment]
Hello everyone
Is this enchantment still viable? I'd be glad to resolve this one.

**Comment by dawagner** [comment]
Hi, @divserge. Is it still relevant and you are welcome to tackle it; thanks!

**Comment by DanielLowry** [comment]
I took a quick look at this, starting at CParameterMgr and working upwards changing the necessary parameters/member variables.  However, I hit an issue when hitting ConfigurableElementAggregator.cpp whereby it defines the following function pointer typedef (in .h):

typedef bool (CConfigurableElement::*MatchesAggregationCriterion)() const;

The issue being that in if the parameter for the doAggregate function is changed from a const * CConfigurableElement to a const std::unique_ptr<CConfigurableElement> then the function pointer dereferencing here:

if(!(pConfigurableElement->*_pfnMatchesAggregationCriterion)())

fails to work (as far as I can tell, because the ->* operator has not been defined properly for smart pointers...yet).

The only other option I can see (though someone with better C++ will probably have something better) is to use shared pointers instead - then for this doAggregate issue in particular, presumably a raw pointer could also be created to the object (which in turn could then be used as in the code right now).

Not ideal, but haven't got any other ideas off the top of my head.

Any thoughts?
